src/uct/ib/dc/dc_mlx5.c

```diff
@@ -1469,15 +1476,19 @@ static UCS_CLASS_INIT_FUNC(uct_dc_mlx5_iface_t, uct_md_h tl_md, uct_worker_h wor
     kh_init_inplace(uct_dc_mlx5_fc_hash, &self->tx.fc_hash);
 
     self->tx.rand_seed = config->rand_seed ? config->rand_seed : time(NULL);
-    self->tx.pend_cb   = uct_dc_mlx5_iface_is_dci_rand(self) ?
+    self->tx.pend_cb   = uct_dc_mlx5_iface_is_dci_shared(self) ?
                          uct_dc_mlx5_iface_dci_do_rand_pending_tx :
                          uct_dc_mlx5_iface_dci_do_dcs_pending_tx;
 
-    /* calculate num_dci_channels: select minimum from requested by runtime
-     * and supported by HCA, must be power of two */
-    num_dci_channels = ucs_roundup_pow2(ucs_max(config->num_dci_channels, 1));
-    self->tx.num_dci_channels = ucs_min(num_dci_channels,
-                                        UCS_BIT(md->log_max_dci_stream_channels));
+    if (tx_policy == UCT_DC_TX_POLICY_HW_DCS) {
+        /* calculate num_dci_channels: select minimum from requested by runtime
+          * and supported by HCA, must be power of two */
+        num_dci_channels = ucs_roundup_pow2(ucs_max(config->num_dci_channels, 1));
```

Comment:
why it has to be power of 2?

Reply:
this is an old comment, I just moved it inside the "if".

Reply:
yes, because when you call create_dci API, you pass log(max_dcis).

Reply:
ok
but why ucs_max is needed? config->num_dci_channels should not be 0